sound/soc/sdca/sdca_asoc.c

```diff
@@ -863,43 +864,160 @@ static int control_limit_kctl(struct device *dev,
 	max = sign_extend32(max, control->nbits - 1);
 
 	/*
-	 * FIXME: Only support power of 2 step sizes as this can be supported
-	 * by a simple shift.
+	 * The SDCA volumes are in steps of 1/256th of a dB, so we need to
+	 * scale them to 1/100ths of a dB for the ALSA TLV.
+	 * The min and max values are in Q7.8 format, so we need to scale
+	 * them to 1/100ths of a dB.
 	 */
-	if (hweight32(step) != 1) {
-		dev_err(dev, "%s: %s: currently unsupported step size\n",
-			entity->label, control->label);
-		return -EINVAL;
-	}
-
-	/*
-	 * The SDCA volumes are in steps of 1/256th of a dB, a step down of
-	 * 64 (shift of 6) gives 1/4dB. 1/4dB is the smallest unit that is also
-	 * representable in the ALSA TLVs which are in 1/100ths of a dB.
-	 */
-	shift = max(ffs(step) - 1, 6);
+	min = min * 100 / SCALE_FACTOR;
+	max = max * 100 / SCALE_FACTOR;
+	step = step * 100 / SCALE_FACTOR;
 
 	tlv = devm_kcalloc(dev, 4, sizeof(*tlv), GFP_KERNEL);
 	if (!tlv)
 		return -ENOMEM;
 
 	tlv[0] = SNDRV_CTL_TLVT_DB_SCALE;
 	tlv[1] = 2 * sizeof(*tlv);
-	tlv[2] = (min * 100) >> 8;
-	tlv[3] = ((1 << shift) * 100) >> 8;
+	tlv[2] = min;
+	tlv[3] = step;
 
-	mc->min = min >> shift;
-	mc->max = max >> shift;
-	mc->shift = shift;
-	mc->rshift = shift;
-	mc->sign_bit = 15 - shift;
+	mc->min = 0;
+	mc->max = (max - min) / step;
```

Comment:
I will need to spend a little more time to work through the maths on these ones. But my initial concern would be with rounding errors, are you relatively sure rounding works out sensible here? I guess ultimately user-space tends to just treat the control as a min to max control anyway so maybe it doesn't matter?

Reply:
Ok... I think mostly this is fine since the number of steps is calculated after the calculation of the step size, so kinda compensates for the error in the step size. However, there is one corner case we probably need to think about which is the case where SDCA gives a step size below 1/100th of a dB this would currently result in a 0dB step size. So we still need to do some sort of capping on the step size. I would be a little tempted to keep the capping to 1/4dB steps but I could be persuaded otherwise if you really wanted a smaller cap.